contracts/message-bus/MessageInbox.sol

```diff
@@ -0,0 +1,232 @@
+
+
+    /* External Functions. */
+
+    /**
+     * @notice Generate inbox message hash from the input params
+     * @param _intentHash Intent hash of message.
+     * @param _nonce Nonce of sender.
+     * @param _gasPrice Gas price.
+     * @param _gasLimit Gas limit.
+     * @param _sender Sender address.
+     * @return messageHash_ Message hash.
+     */
+    function inboxMessageHash(
+        bytes32 _intentHash,
+        uint256 _nonce,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        address _sender
+    )
+        external
+        view
+        returns (bytes32 messageHash_)
+    {
+        messageHash_ = _messageHash(
+            _intentHash,
+            _nonce,
+            _gasPrice,
+            _gasLimit,
+            _sender,
+            inboundMessageIdentifier
+        );
+    }
+
+
+    /* Internal Functions. */
+
+    /**
+     * @notice Setup message inbox.
+     *
+     * @dev Function requires:
+     *          - inboundMessageIdentifier must be zero
+     *          - metachainId must not be zero
+     *          - messageOutbox address must not be zero
+     *          - verifyingAddress must not be zero
+     *          - stateRootProvider must not be zero
+     *
+     * @param _metachainId Metachain identifier.
+     * @param _messageOutbox MessageOutbox contract address.
+     * @param _outboxStorageIndex Storage index of outbox mapping in
+     *                            MessageOutbox contract.
+     * @param _stateRootProvider State root provider contract address.
+     * @param _maxStorageRootItems Defines how many storage roots should be
+     *                             stored in circular buffer.
+     * @param _verifyingAddress Address of verifying contract.
+     */
+    function setupMessageInbox(
+        bytes32 _metachainId,
+        address _messageOutbox,
+        uint8 _outboxStorageIndex,
+        StateRootI _stateRootProvider,
+        uint256 _maxStorageRootItems,
+        address _verifyingAddress
+    )
+        internal
+    {
+        require(
+            inboundMessageIdentifier == bytes32(0),
+            "Message inbox is already setup."
+        );
+
+        require(
+            _metachainId != bytes32(0),
+            "metachain id is 0."
+        );
+
+        require(
+            _messageOutbox != address(0),
+            "Inbox address is 0."
+        );
+
+        require(
+            address(_stateRootProvider) != address(0),
+            "State root provider address is 0."
+        );
+
+        require(
+            _verifyingAddress != address(0),
+            "Verifying address is 0."
+        );
+
+        messageOutbox = _messageOutbox;
+        outboxStorageIndex = _outboxStorageIndex;
+
+        inboundMessageIdentifier = keccak256(
+            abi.encode(
+                DOMAIN_SEPARATOR_TYPEHASH,
+                DOMAIN_SEPARATOR_NAME,
+                DOMAIN_SEPARATOR_VERSION,
+                _metachainId,
+                _verifyingAddress
```

Comment:
will this be `address(this)` 

EIP712 doc says
> address verifyingContract the address of the contract that will verify the signature. The user-agent may do contract specific phishing prevention.

https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator

Reply:
The Gateways that implement this contract, will call with the proper address.